### PR TITLE
adding stepper motor example

### PR DESCRIPTION
**Pull request by benschw**
Control a stepper motor (28BJY-48)

(this is a port of Matt Hawkins' example impl from http://www.raspberrypi-spy.co.uk/2012/07/stepper-motor-control-in-python/)

Datasheet:
http://www.raspberrypi-spy.co.uk/wp-content/uploads/2012/07/Stepper-Motor-28BJY-48-Datasheet.pdf

### CODE REVIEW

File: samples/28bjy-48.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+//
+// Datasheet:
+// http://www.raspberrypi-spy.co.uk/wp-content/uploads/2012/07/Stepper-Motor-28BJY-48-Datasheet.pdf
+//
+// this is a port of Matt Hawkins' example impl from
+// http://www.raspberrypi-spy.co.uk/2012/07/stepper-motor-control-in-python/
+// (link privides additional instructions for wiring your pi)
+
+import (
+	"flag"
```
Comment (kidoman) [inline samples/28bjy-48.go]: Please run goimports

File: samples/28bjy-48.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+	stepCount := len(seq) - 1
+	stepDir := 2 // Set to 1 or 2 for clockwise, -1 or -2 for counter-clockwise
+
+	// Start main loop
+	quit := make(chan os.Signal, 1)
+	signal.Notify(quit, os.Interrupt, os.Kill)
+	defer signal.Stop(quit)
+
+	stepCounter := 0
```
Comment (kidoman) [inline samples/28bjy-48.go]:
```
var stepCounter int
```
is more idiomatic

File: samples/28bjy-48.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	for {
+		select {
+		case <-time.After(time.Duration(*stepWait) * time.Millisecond):
```
Comment (kidoman) [inline samples/28bjy-48.go]: Would a ```time.Ticker``` be more apt for this usage?

**Comment by kidoman** [comment]
Thanks for the PR. Please have a look at some of the review comments

File: samples/28bjy-48.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	"fmt"
+	"github.com/kidoman/embd"
+	_ "github.com/kidoman/embd/host/rpi"
+	"os"
+	"os/signal"
+	"time"
+)
+
+func main() {
+	stepWait := flag.Int("step-delay", 10, "milliseconds between steps")
```
Comment (kunalpowar) [inline samples/28bjy-48.go]: stepWait can be stepDelay, to be consistent with step-delay flag.

Comment (kidoman) [inline samples/28bjy-48.go]: +1

**Comment by kunalpowar** [comment]
Thanks for the PR @benschw. I'm thinking if this can be pulled out into a package which exposes something like
```
stepper, _ := embd.NewStepper(...)
```

**Comment by benschw** [comment]
thanks for all the feedback! I'll try to tackle the updates as soon as possible

**Comment by benschw** [comment]
I've incorporated your feedback, except the idea of creating a `Stepper` struct. If you'd like me to try that, maybe getting this approved and creating a new PR pulling the Stepper abstraction out and refactoring the sample to use it would make review easier.

Let me know how to proceed.

**Comment by kidoman** [comment]
Sounds about right

**Comment by benschw** [comment]
sorry - so should I cancel this pr? or do you want to pull it before I get started on the Stepper struct?

**Comment by kidoman** [comment]
No I am going to merge this. That should be a second PR

File: samples/28bjy-48.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	stepCount := len(seq) - 1
+	stepDir := 2 // Set to 1 or 2 for clockwise, -1 or -2 for counter-clockwise
+
+	quit := make(chan os.Signal, 1)
+	signal.Notify(quit, os.Interrupt, os.Kill)
+	defer signal.Stop(quit)
+
+	// Start main loop
+	ticker := time.NewTicker(time.Duration(*stepDelay) * time.Millisecond)
+
```
Comment (kidoman) [inline samples/28bjy-48.go]:
```
defer timer.Close()
```

File: samples/28bjy-48.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+	var stepCounter int
+	for {
+		select {
+		case <-ticker.C:
+
```
Comment (kidoman) [inline samples/28bjy-48.go]: This empty line is unnecessary

File: samples/28bjy-48.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+			}
+			stepCounter += stepDir
+
+			// If we reach the end of the sequence start again
+			if stepCounter >= stepCount {
+				stepCounter = 0
+			} else if stepCounter < 0 {
+				stepCounter = stepCount
+			}
+
```
Comment (kidoman) [inline samples/28bjy-48.go]: This empty line is good +1

File: samples/28bjy-48.go [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+		defer pin.Close()
+		if err := pin.SetDirection(embd.Out); err != nil {
+			panic(err)
+		}
+		if err := pin.Write(embd.Low); err != nil {
+			panic(err)
+		}
+		defer func() {
+			if err := pin.SetDirection(embd.In); err != nil {
+				panic(err)
```
Comment (kidoman) [inline samples/28bjy-48.go]: I would not bother panicing during a defer.

```
defer pin.SetDirection(embd.In)
```

would be more idiomatic.

File: samples/28bjy-48.go [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+				if seq[stepCounter][i] != 0 {
+					fmt.Printf("Enable pin %d, step %d\n", i, stepCounter)
+					if err := pin.Write(embd.High); err != nil {
+						panic(err)
+					}
+				} else {
+					if err := pin.Write(embd.Low); err != nil {
+						panic(err)
+					}
+
```
Comment (kidoman) [inline samples/28bjy-48.go]: This empty line is unnecessary

**Comment by benschw** [comment]
Sorry, I wasn't totally clear on the empty lines. Could you let me know line numbers if I got any wrong?

File: samples/28bjy-48.go [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+	stepCount := len(seq) - 1
+	stepDir := 2 // Set to 1 or 2 for clockwise, -1 or -2 for counter-clockwise
+
+	quit := make(chan os.Signal, 1)
+	signal.Notify(quit, os.Interrupt, os.Kill)
+	defer signal.Stop(quit)
+
+	// Start main loop
+	ticker := time.NewTicker(time.Duration(*stepDelay) * time.Millisecond)
+	defer timer.Close()
```
Comment (kidoman) [inline samples/28bjy-48.go]: Should be ticker.Close()

**Comment by kidoman** [comment]
About the empty lines: I have left line comments whereever I noticed an issue

**Comment by benschw** [comment]
oops, I'm a little embarrassed about not running that before pushing.

**Comment by benschw** [comment]
also, I moved ticker.Stop up to a defer defined when i start the ticker; there is no `Close` implemented on `Ticker`

**Pull request merged**